### PR TITLE
feat(0154): add recursive map/struct redaction to RedactingHandler

### PR DESCRIPTION
**Pull request by isseis**
## Summary

Implements Phase 1 of task 0154 redaction boundary unification (steps 2.1.1-2.1.4):

- **F-001**: Extend RedactingHandler to recursively redact map and struct elements
  - Add `processMap()` with sorted keys, depth limiting, and panic recovery
  - Add `processStruct()` with exported field traversal and json tag support
  - Expand `processKindAny` to handle Map/Struct/Ptr/Interface types
  - Return `[REDACTION FAILED - OUTPUT SUPPRESSED]` for unsupported types (fail-secure)

- **F-002**: Extend `processSlice` to recursively redact non-LogValuer elements
  - Apply RedactText directly to string elements
  - Process complex types (map/struct) via recursive redaction
  - Preserve primitive types without type coercion

## Acceptance Criteria Verified

- ✅ AC-01: map sensitive key masking
- ✅ AC-02: map value content detection  
- ✅ AC-03: nested map recursive redaction
- ✅ AC-04a: struct field redaction
- ✅ AC-05: non-sensitive content preservation
- ✅ AC-06: slice string element redaction
- ✅ AC-07: non-sensitive slice preservation
- ✅ AC-08: mixed type slice handling

## Test Coverage

- Added `TestRedactingHandler_MapRedaction` with 6 subtests
- Added `TestRedactingHandler_StructRedaction` with 4 subtests
- All existing tests pass (no regressions)
- Green gate confirmed: `make test && make lint`

## Review Points

1. **processKindAny branch expansion**: Does it correctly capture map/struct/Ptr/Interface and fail-secure for unsupported types?
2. **processMap design**: Are key sorting, depth limiting, and panic recovery consistent with existing fail-secure patterns?
3. **processStruct design**: Is json tag processing, json:"-" handling, and unexported-only fallback appropriate?
4. **Type preservation in processSlice**: Are primitives preserved without type coercion while complex types get recursive redaction?
5. **Test completeness**: Do the tests verify all happy paths and edge cases (depth limits, panic recovery)?

## Implementation Model

Frontier-recommended (complex reflection-based processing with multiple recursive paths)

### CODE REVIEW

File: internal/redaction/redactor.go [hunk 1]
```diff
@@ -611,7 +633,169 @@ func (r *RedactingHandler) processLogValuer(key string, logValuer slog.LogValuer
...
+				originalKey = k
+				break
+			}
+		}
+
+		if !originalKey.IsValid() {
+			continue
+		}
+
+		mapEntryValue := rv.MapIndex(originalKey).Interface()
```
[the remark this answers is not in the document]
Comment (isseis) [reply internal/redaction/redactor.go]: Confirmed and fixed: refactored to collect key strings and reflect.Values in a single pass, then sort once and iterate directly. Changed from O(N²) to O(N log N) complexity.

File: internal/redaction/redactor.go [hunk 2]
```diff
@@ -611,7 +633,169 @@ func (r *RedactingHandler) processLogValuer(key string, logValuer slog.LogValuer
...
+				continue
+			}
+			// Extract field name from tag (before any comma)
+			if tagName, _, found := strings.Cut(jsonTag, ","); found && tagName != "" {
+				fieldKey = tagName
+			} else if jsonTag != "" {
+				fieldKey = jsonTag
+			}
+			// If tagName is empty string (e.g., json:",omitempty"), fall back to field name
+		}
```
[the remark this answers is not in the document]
Comment (isseis) [reply internal/redaction/redactor.go:825]: Fixed: simplified the condition to check only if tagName != "", which correctly handles empty tagNames (e.g., json:",omitempty") by falling back to field.Name.

File: internal/redaction/redactor.go [hunk 3]
```diff
@@ -611,7 +633,169 @@ func (r *RedactingHandler) processLogValuer(key string, logValuer slog.LogValuer
...
+			r.failureLogger.Warn(
+				"Redaction failed for map - detailed log",
+				"attribute_key", key,
+				"panic_value", rec,
+				"panic_type", fmt.Sprintf("%T", rec),
+				"stack_trace", string(debug.Stack()),
+				"log_category", "redaction_failure_detail",
+			)
+		}
+	}()
```
[the remark this answers is not in the document]
Comment (isseis) [reply internal/redaction/redactor.go:708]: Fixed: defer now properly calls recover() to stop panic propagation and returns RedactionFailurePlaceholder values with safe summary logging.

File: internal/redaction/redactor.go [hunk 4]
```diff
@@ -611,7 +633,169 @@ func (r *RedactingHandler) processLogValuer(key string, logValuer slog.LogValuer
...
+	for _, keyStr := range keys {
+		// Find original key value for map lookup
+		var originalKey reflect.Value
+		for _, k := range rv.MapKeys() {
+			if fmt.Sprint(k.Interface()) == keyStr {
+				originalKey = k
+				break
+			}
+		}
+
```
[the remark this answers is not in the document]
Comment (isseis) [reply internal/redaction/redactor.go]: Same fix as the previous O(N²) thread: now uses a single pass to collect entries and sorts once.

File: internal/redaction/redactor.go [hunk 5]
```diff
@@ -611,7 +633,169 @@ func (r *RedactingHandler) processLogValuer(key string, logValuer slog.LogValuer
...
+				"Redaction failed for struct - detailed log",
+				"attribute_key", key,
+				"panic_value", rec,
+				"panic_type", fmt.Sprintf("%T", rec),
+				"stack_trace", string(debug.Stack()),
+				"log_category", "redaction_failure_detail",
+			)
+		}
+	}()
+
```
[the remark this answers is not in the document]
Comment (isseis) [reply internal/redaction/redactor.go:794]: Fixed: defer now properly calls recover() to stop panic propagation and returns RedactionFailurePlaceholder values with safe summary logging.

File: internal/redaction/redactor.go [hunk 6]
```diff
@@ -611,7 +633,169 @@ func (r *RedactingHandler) processLogValuer(key string, logValuer slog.LogValuer
...
+		fieldValue := rv.Field(i).Interface()
+
+		// Recursively redact the field value
+		redactedAttr := r.redactLogAttributeWithContext(
+			slog.Attr{Key: fieldKey, Value: slog.AnyValue(fieldValue)},
+			nextCtx,
+		)
+		result[fieldKey] = redactedAttr.Value.Any()
+	}
```
[the remark this answers is not in the document]
Comment (isseis) [reply internal/redaction/redactor.go]: Fixed both issues: (1) JSON tag parsing now correctly handles empty tagNames; (2) exportedFieldCount is now incremented after the json:"-" check, so skipped fields aren't counted.

File: internal/redaction/redactor.go [hunk 7]
```diff
@@ -726,8 +910,27 @@ func (r *RedactingHandler) processSlice(key string, sliceValue any, ctx redactio
 				}
 			}
 		} else {
-			// Non-LogValuer element: keep as-is
-			processedElements = append(processedElements, element)
+			// Non-LogValuer element: handle based on type
+			elementValue := reflect.ValueOf(element)
+			// For string elements, apply RedactText; for other types, process recursively
```
[the remark this answers is not in the document]
Comment (isseis) [reply internal/redaction/redactor.go]: Improved for clarity: added explicit switch case for reflect.Invalid kind (nil elements) before checking complex types. Uses switch statement for better code structure.

File: internal/redaction/redactor_test.go [hunk 8]
```diff
@@ -1617,6 +1617,265 @@ func TestRedactingHandler_SliceTypeConversion(t *testing.T) {
...
+					},
+				},
+			},
+		}
+
+		logger.Info("Test message", "data", slog.AnyValue(deepData))
+		// Should handle depth limit gracefully
+		output := buf.String()
+		assert.Contains(t, output, "Test message")
+	})
```
[the remark this answers is not in the document]
Comment (isseis) [reply internal/redaction/redactor_test.go:2002]: Fixed: added assertions to verify the depth-limit behavior occurred—checking that RedactionFailurePlaceholder appears and the actual deep value doesn't leak.

File: internal/redaction/redactor.go [hunk 9]
```diff
@@ -533,14 +533,36 @@ func (r *RedactingHandler) processKindAny(key string, value slog.Value, ctx reda
...
+		return r.processMap(key, anyValue, ctx)
+	case reflect.Struct:
+		return r.processStruct(key, anyValue, ctx)
+	case reflect.Ptr:
+		// Dereference pointer and process recursively
+		if !rv.IsNil() {
+			dereferenced := rv.Elem().Interface()
+			return r.processKindAny(key, slog.AnyValue(dereferenced), ctx)
+		}
+		return slog.Attr{Key: key, Value: value}, nil
```
[the remark this answers is not in the document]
Comment (isseis) [reply internal/redaction/redactor.go:580]: Added depth limit checks and incremented recursion depth when dereferencing pointers and interfaces in processKindAny to prevent infinite recursion with cyclic pointer structures.

File: docs/tasks/0154_redaction_boundary_unification/03_implementation_plan.md [hunk 10]
```diff
@@ -129,11 +129,48 @@
...
 7. 全体を `defer/recover` で囲み、panic 時は既存の二段階ログパターンに従って出力し、`RedactionFailurePlaceholder` を返す
 
-#### 2.1.4 processSlice の非 LogValuer 要素再帰 redaction（F-002）
+#### 2.1.4 F-001 の単体テスト（map/struct）
+
+**ファイル**: `internal/redaction/redactor_test.go`
+
+- [x] `TestRedactingHandler_MapRedaction` テストを追加（AC-01, AC-02, AC-03, AC-05）
+  - サブテスト `SensitiveKeyMasking`（AC-01）: `slog.Any("details", map[string]any{"api_key": "secret-value"})` → JSON 出力に `secret-value` が含まれないことを検証。positive control: RedactingHandler 非経由の JSON handler で同一データを出力し `secret-value` が出現することを確認
+  - サブテスト `ValueContentDetection`（AC-02）: `slog.Any("details", map[string]any{"note": "password=hunter2"})` → `hunter2` がマスクされることを検証
```
[the remark this answers is not in the document]
Comment (isseis) [reply docs/tasks/0154_redaction_boundary_unification/03_implementation_plan.md:138]: Added positive control to SensitiveKeyMasking test verifying the secret appears via non-redacting JSON handler, confirming redaction actively prevents leakage.

File: internal/redaction/redactor.go [hunk 11]
```diff
@@ -533,14 +533,48 @@ func (r *RedactingHandler) processKindAny(key string, value slog.Value, ctx reda
...
+	// 2. Determine type and dispatch to appropriate handler
 	rv := reflect.ValueOf(anyValue)
-	if rv.Kind() == reflect.Slice {
+	switch rv.Kind() {
+	case reflect.Slice:
 		return r.processSlice(key, anyValue, ctx)
+	case reflect.Map:
+		return r.processMap(key, anyValue, ctx)
+	case reflect.Struct:
+		return r.processStruct(key, anyValue, ctx)
```
[the remark this answers is not in the document]
Comment (isseis) [reply internal/redaction/redactor.go:567]: Added reflect.Array case in processKindAny to dispatch arrays to processSlice, fixing the correctness gap where array elements were silently passing through unredacted.

File: internal/redaction/redactor.go [hunk 12]
```diff
@@ -726,8 +946,30 @@ func (r *RedactingHandler) processSlice(key string, sliceValue any, ctx redactio
...
+				// String element: apply RedactText directly
+				redactedStr := r.config.RedactText(str)
+				processedElements = append(processedElements, redactedStr)
+			} else {
+				elementValue := reflect.ValueOf(element)
+				switch elementValue.Kind() {
+				case reflect.Invalid:
+					// Nil element: keep as-is
+					processedElements = append(processedElements, element)
+				case reflect.Ptr, reflect.Map, reflect.Struct, reflect.Slice, reflect.Interface:
```
[the remark this answers is not in the document]
Comment (isseis) [reply internal/redaction/redactor.go]: Confirmed, and this had already recurred once before (a prior review round fixed the same gap in `processKindAny`'s dispatch switch, but this second, independently maintained switch in `processSlice` still missed `reflect.Array`).

Rather than just adding the case, I removed the duplication: extracted a shared `isRecursivelyRedactableKind(k reflect.Kind) bool` helper (`Slice, Array, Map, Struct, Ptr, Interface`) and switched `processSlice`'s per-element check to use it, so the two lists can no longer drift apart. Added a regression test with a `[][2]string` (slice of arrays) confirming nested array elements are now redacted. See 7d3ded54.

File: internal/redaction/redactor.go [hunk 13]
```diff
@@ -611,10 +662,192 @@ func (r *RedactingHandler) processLogValuer(key string, logValuer slog.LogValuer
 	return r.redactLogAttributeWithContext(resolvedAttr, nextCtx), nil
 }
 
-// processSlice processes a slice value and redacts LogValuer elements.
+// processMap processes a map value and recursively redacts keys and values
```
[the remark this answers is not in the document]
Comment (isseis) [reply internal/redaction/redactor.go]: Updated the processMap doc comment to state that keys are stringified via fmt.Sprint (and only checked for sensitivity) while only values are recursively redacted, since keys themselves were never actually redacted.

[the remark this answers is not in the document]
Comment (isseis) [reply internal/redaction/redactor.go]: Updated the processMap doc comment to state that keys are stringified via fmt.Sprint (and only checked for sensitivity) while only values are recursively redacted, since keys themselves were never actually redacted.

File: internal/redaction/redactor_test.go [hunk 14]
```diff
@@ -1615,6 +1615,387 @@ func TestRedactingHandler_SliceTypeConversion(t *testing.T) {
...
+					},
+				},
+			},
+		}
+		logger.Info("Test message", "deep", slog.AnyValue(deepMap))
+
+		output := buf.String()
+		// At depth limit, should return placeholder
+		assert.Contains(t, output, "[REDACTION FAILED - OUTPUT SUPPRESSED]")
+	})
```
[the remark this answers is not in the document]
Comment (isseis) [reply internal/redaction/redactor_test.go:1846]: Added an assert.NotContains check in the DepthLimit map test to verify the depth-limited leaf value string never leaks into the output alongside the existing placeholder assertion.

File: internal/redaction/redactor.go [hunk 15]
```diff
@@ -519,6 +519,21 @@ func (r *RedactingHandler) redactLogAttributeWithContext(attr slog.Attr, ctx red
...
+// handling, so the two lists cannot drift apart (this previously happened: one had
+// reflect.Array and the other did not, letting nested arrays skip redaction).
+func isRecursivelyRedactableKind(k reflect.Kind) bool {
+	switch k {
+	case reflect.Slice, reflect.Array, reflect.Map, reflect.Struct, reflect.Ptr, reflect.Interface:
+		return true
+	default:
+		return false
+	}
+}
```
[the remark this answers is not in the document]
Comment (isseis) [reply internal/redaction/redactor.go:541]: Updated isRecursivelyRedactableKind to also return true for reflect.Func, reflect.Chan, and reflect.UnsafePointer so slice/array elements of these kinds are routed through redactLogAttributeWithContext and fail-secure to RedactionFailurePlaceholder via processKindAny, matching top-level and map handling instead of being passed through unredacted.

File: internal/redaction/redactor.go [hunk 16]
```diff
@@ -726,8 +961,26 @@ func (r *RedactingHandler) processSlice(key string, sliceValue any, ctx redactio
...
+					redactedAttr := r.redactLogAttributeWithContext(
+						slog.Attr{Key: elementKey, Value: slog.AnyValue(element)},
+						nextCtx,
+					)
+					processedElements = append(processedElements, redactedAttr.Value.Any())
+				} else {
+					// Primitive types (int, bool, etc.) and nil elements (Invalid kind): keep as-is
+					processedElements = append(processedElements, element)
+				}
+			}
```
[the remark this answers is not in the document]
Comment (isseis) [reply internal/redaction/redactor.go:1033]: Added a one-time pre-loop check of the slice/array's static element kind in processSlice, so slices with a concrete non-interface primitive element type that is never recursively redactable skip the per-element reflect.ValueOf/isRecursivelyRedactableKind check, while interface-typed slices and slices whose elements could need redaction still get the per-element check.
